infra/Pos/Network/Types.hs

```diff
@@ -0,0 +1,171 @@
+module Pos.Network.Types
+    ( NetworkConfig (..)
+    , Topology(..)
+    , SubscriptionWorker(..)
+    , Valency
+    , Fallbacks
+    , topologyNodeType
+    , topologySubscriberNodeType
+    , topologySubscriptionWorker
+    , topologyRunKademlia
+    , resolveDnsDomains
+    , defaultNetworkConfig
+    , staticallyKnownPeers
+    , initQueue
+      -- * Re-exports
+      -- ** from .DnsDomains
+    , DnsDomains(..)
+    , DNSError
+      -- ** from time-warp
+    , NodeType (..)
+    , MsgType (..)
+    , Origin (..)
+      -- ** other
+    , NodeId (..)
+    ) where
+
+import           Universum
+import           Data.IP (IPv4)
+import           Network.Broadcast.OutboundQueue.Types
+import           Network.Broadcast.OutboundQueue (OutboundQ)
+import qualified Network.Broadcast.OutboundQueue as OQ
+import           Node.Internal (NodeId (..))
+import           Pos.Network.DnsDomains (DnsDomains(..), DNSError)
+import           Pos.Network.Yaml (NodeName(..))
+import           Pos.Util.TimeWarp  (addressToNodeId)
+import qualified Pos.Network.DnsDomains as DnsDomains
+import qualified Data.ByteString.Char8  as BS.C8
+
+-- | Information about the network in which a node participates.
+data NetworkConfig kademlia = NetworkConfig
+    { ncTopology :: !(Topology kademlia)
+      -- ^ Network topology from the point of view of the current node
+    , ncDefaultPort :: !Word16
+      -- ^ Port number to use when translating IP addresses to NodeIds
+    , ncSelfName :: !(Maybe NodeName)
+      -- ^ Our node name (if known)
+    }
+  deriving (Show)
+
+defaultNetworkConfig :: Topology topology -> NetworkConfig topology
+defaultNetworkConfig ncTopology = NetworkConfig {
+      ncDefaultPort = 3000
+    , ncSelfName    = Nothing
+    , ..
+    }
+
+type Valency = Int
+
+type Fallbacks = Int
+
+-- | Topology of the network, from the point of view of the current node
+data Topology kademlia =
+    -- | All peers of the node have been statically configured
+    --
+    -- This is used for core and relay nodes
+    TopologyCore !(Peers NodeId)
+
+  | TopologyRelay !(Peers NodeId) !kademlia
+
+    -- | We discover our peers through DNS
+    --
+    -- This is used for behind-NAT nodes.
+  | TopologyBehindNAT !DnsDomains
+
+    -- | We discover our peers through Kademlia
+  | TopologyP2P !Valency !Fallbacks !kademlia
+
+    -- | We discover our peers through Kademlia, and every node in the network
+    -- is a core node.
+  | TopologyTraditional !Valency !Fallbacks !kademlia
+
+    -- | Light wallets simulate "real" edge nodes, but are configured with
+    -- a static set of relays.
+  | TopologyLightWallet ![NodeId]
+  deriving (Show)
+
```

Comment:
This looks great, along with the topology interpretation functions below than centralises policy here.